### PR TITLE
Reconcile v0.10 with new CI

### PR DESCRIPTION
**Pull request by orangemocha**
As part of CI reconciliation, these commits make v0.10 work with the jobs at https://jenkins-iojs.nodesource.com/

The first two commits (https://github.com/joyent/node/commit/9ddc686d41971a90de9a43a21644fc8794c679d8 and https://github.com/joyent/node/commit/851e32c994dc671c26c242f2f476c91b507ccd81) are probably not strictly necessary, but they do a useful refactoring and they bring the v0.10 test runner closer to the v0.12 version.

The last commit (https://github.com/joyent/node/commit/783cc5e97b27ed78e014a098fb541362353e27fb) is an attempt to mark the current test failures in v0.10 as flaky (there a quite a few!). I will file issues to investigate those as soon as this PR lands. Some of these tests fail every time, so they are not actually flaky. A platform-specific skip might be more appropriate for those, but I defer that decision to the investigation that will follow, so that we can unblock node-accept-pull-request and CI reconciliation.

The other commits in this PR are pretty much a backport of https://github.com/joyent/node/pull/25653 to v0.10.

/cc @misterdjules @rvagg @jbergstroem

### CODE REVIEW

File: tools/test.py [hunk 1]
```diff
@@ -1452,15 +1487,15 @@ def wrap(processor):
 
   result = None
   def DoSkip(case):
-    return SKIP in case.outcomes or SLOW in case.outcomes
+    return SKIP in case.outcomes or SLOW in case.outcomes or (FLAKY in case.outcomes and options.flaky_tests == "skip")
```
Comment (misterdjules) [inline tools/test.py]: Styling: can we put the added condition on another line to avoid that long line?

File: test/simple/simple.status [hunk 2]
```diff
@@ -1 +1,35 @@
 prefix simple
+
+test-http-client-timeout-event          : PASS,FLAKY
```
Comment (misterdjules) [inline test/simple/simple.status]: @orangemocha How did you gather that information? By looking at the history of failing tests for the nodejs-v0.10 Jenkins job on jenkins.nodejs.org?

**Comment by misterdjules** [comment]
Thank you for doing that :+1: Ran `make test` in my local v0.10 clone with these commits and all was fine.

So aside from the style nitpick and my question, LGTM!

**Comment by misterdjules** [comment]
@orangemocha Also, if you haven't used it in the past, remember to use https://www.npmjs.com/package/git-apply-pr to merge the change in v0.10.

File: tools/test.py (hunk 1 above)
Comment (orangemocha) [inline tools/test.py]: Sure

File: test/simple/simple.status (hunk 2 above)
Comment (orangemocha) [inline test/simple/simple.status]: I looked at recent logs and I also ran the job in the new Jenkins a few times collecting all the failures.

**Comment by orangemocha** [comment]
>  Also, if you haven't used it in the past, remember to use https://www.npmjs.com/package/git-apply-pr to merge the change in v0.10.

I can now use node-accept-pull-request to land this :smile:

**Comment by orangemocha** [comment]
Fixed according to feedback.

File: test/simple/simple.status (hunk 2 above)
Comment (misterdjules) [inline test/simple/simple.status]: Sounds good, thanks!

**Comment by misterdjules** [comment]
@orangemocha 

> > Also, if you haven't used it in the past, remember to use https://www.npmjs.com/package/git-apply-pr to merge the change in v0.10.

> I can now use node-accept-pull-request to land this 

Ah great, I thought some of the actually failing and non-flaky tests had been left out of the list of flaky tests, but it makes sense to mark them as flaky. Excellent :+1:

**Comment by orangemocha** [comment]
Landed in https://github.com/joyent/node/commit/a2f879f197f76d91a42ec0f8322f18948911a552 and adjacent commits.

Merge courtesy of https://jenkins-iojs.nodesource.com/job/node-accept-pull-request/21/ :smile:

**Comment by orangemocha** [comment]
.. and filed issues for all the tests marked as flaky, as you might have seen in your inbox (or spam folder :wink: )

**Comment by misterdjules** [comment]
@orangemocha Thank you for taking the time to do that :+1:

**Pull request closed without merging**